File: apps/runtime/src/git/access.rs
//! Application-level Git execution policy, not an operating-system sandbox.
//! Worktree inspection may invoke clean filters even for `status`. Without
//! execution permission only built-in metadata/index/object reads are admitted;
//! no conversion is silently removed before a write.
use crate::error::{AppError, AppResult};

pub fn require_execution(allowed: bool, purpose: &str) -> AppResult<()> {
    if allowed {
        Ok(())
    } else {
        Err(AppError::GitExecutionRequired(format!(
            "{purpose} requires workspace execution permission: Git can run repository filters, hooks, or transport helpers"
        )))
    }
}

/// Prefix only commands whose arguments are built by our typed Git services.
/// This function is deliberately not a general-purpose safe-command evaluator.
pub fn arguments(arguments: Vec<String>, allow_helpers: bool) -> AppResult<Vec<String>> {
    if allow_helpers {
        return Ok(arguments);
    }
    let verb = arguments.first().map(String::as_str).unwrap_or("");
    let positional = arguments
        .iter()
        .skip(1)
        .take_while(|arg| arg.as_str() != "--")
        .filter(|arg| !arg.starts_with('-'))
        .collect::<Vec<_>>();
    let immutable_diff = arguments.iter().any(|arg| arg == "--cached")
        || (positional.len() == 2 && positional.iter().all(|arg| is_oid(arg)));
    let allowed = match verb {
        "rev-parse" | "for-each-ref" | "rev-list" | "log" | "show" | "cat-file"
        | "check-ref-format" | "ls-files" => true,
        "symbolic-ref" => positional.len() == 1,
        "remote" => arguments.len() == 1,
        "stash" => arguments.get(1).is_some_and(|arg| arg == "list"),
        "diff" => immutable_diff,
        "diff-tree" => positional.len() == 1 && is_oid(positional[0]),
        _ => false,
    };
    if !allowed
        || arguments
            .iter()
            .take_while(|arg| arg.as_str() != "--")
            .any(|arg| {
                matches!(
                    arg.as_str(),
                    "--filters"
                        | "--textconv"
                        | "--ext-diff"
                        | "--show-signature"
                        | "--alternate-refs"
                )
            })
    {
        require_execution(false, "This Git inspection")?;
    }
    let mut result = vec![
        "-c".into(),
        "core.fsmonitor=false".into(),
        "-c".into(),
        format!(
            "core.hooksPath={}",
            if cfg!(windows) { "NUL" } else { "/dev/null" }
        ),
        "-c".into(),
        "log.showSignature=false".into(),
        "-c".into(),
        "diff.submodule=short".into(),
    ];
    if verb == "stash" {
        result.extend([
            "stash".into(),
            "list".into(),
            "--no-patch".into(),
            "--no-ext-diff".into(),
            "--no-textconv".into(),
        ]);
        result.extend(arguments.into_iter().skip(2));
        return Ok(result);
    }
    result.push(verb.into());
    if matches!(verb, "diff" | "diff-tree" | "log" | "show") {
        result.extend(["--no-ext-diff".into(), "--no-textconv".into()]);
    }
    result.extend(arguments.into_iter().skip(1));
    Ok(result)
}
fn is_oid(value: &str) -> bool {
    matches!(value.len(), 40 | 64) && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

pub fn restrict_sync(command: &mut std::process::Command) {
    // The protocol allow-list overrides per-protocol configuration, including
    // arbitrary remote helpers. Also prevent optional promisor-object fetches.
    command
        .env("GIT_ALLOW_PROTOCOL", "")
        .env("GIT_NO_LAZY_FETCH", "1")
        .env_remove("GIT_EXEC_PATH");
}
pub fn restrict_async(command: &mut tokio::process::Command) {
    command
        .env("GIT_ALLOW_PROTOCOL", "")
        .env("GIT_NO_LAZY_FETCH", "1")
        .env_remove("GIT_EXEC_PATH");
}

File: apps/desktop/src/core/git/access.ts
import { requireExecution, validOid } from "./support";

/**
 * Application-level Git execution policy, not an operating-system sandbox.
 *
 * A direct port of the pre-merge implementation. Worktree inspection may
 * invoke clean filters even for `status`, so without the workspace's execution
 * grant only built-in metadata / index / object reads are admitted, and every
 * admitted command additionally runs with fsmonitor, hooks, signature checks
 * and submodule diffing configured off.
 *
 * This is deliberately **not** a general-purpose safe-command evaluator: it
 * only ever sees argument lists built by the typed Git services in this
 * directory.
 */

const REFUSED_FLAGS = new Set([
  "--filters",
  "--textconv",
  "--ext-diff",
  "--show-signature",
  "--alternate-refs",
]);

export function gitArguments(
  args: readonly string[],
  allowHelpers: boolean,
): string[] {
  if (allowHelpers) return [...args];
  const verb = args[0] ?? "";
  const positional: string[] = [];
  for (const arg of args.slice(1)) {
    if (arg === "--") break;
    if (!arg.startsWith("-")) positional.push(arg);
  }
  const immutableDiff =
    args.includes("--cached") ||
    (positional.length === 2 && positional.every((arg) => validOid(arg)));
  let allowed: boolean;
  switch (verb) {
    case "rev-parse":
    case "for-each-ref":
    case "rev-list":
    case "log":
    case "show":
    case "cat-file":
    case "check-ref-format":
    case "ls-files":
      allowed = true;
      break;
    case "symbolic-ref":
      allowed = positional.length === 1;
      break;
    case "remote":
      allowed = args.length === 1;
      break;
    case "stash":
      allowed = args[1] === "list";
      break;
    case "diff":
      allowed = immutableDiff;
      break;
    case "diff-tree":
      allowed = positional.length === 1 && validOid(positional[0] as string);
      break;
    default:
      allowed = false;
  }
  const beforeSeparator: string[] = [];
  for (const arg of args) {
    if (arg === "--") break;
    beforeSeparator.push(arg);
  }
  if (!allowed || beforeSeparator.some((arg) => REFUSED_FLAGS.has(arg))) {
    requireExecution(false, "This Git inspection");
  }
  const result = [
    "-c",
    "core.fsmonitor=false",
    "-c",
    `core.hooksPath=${process.platform === "win32" ? "NUL" : "/dev/null"}`,
    "-c",
    "log.showSignature=false",
    "-c",
    "diff.submodule=short",
  ];
  if (verb === "stash") {
    result.push(
      "stash",
      "list",
      "--no-patch",
      "--no-ext-diff",
      "--no-textconv",
      ...args.slice(2),
    );
    return result;
  }
  result.push(verb);
  if (["diff", "diff-tree", "log", "show"].includes(verb)) {
    result.push("--no-ext-diff", "--no-textconv");
  }
  result.push(...args.slice(1));
  return result;
}

/**
 * The protocol allow-list overrides per-protocol configuration, including
 * arbitrary remote helpers, and lazy promisor fetches are refused outright.
 */
export function restrictedEnvironment(): Record<string, string | undefined> {
  return {
    GIT_ALLOW_PROTOCOL: "",
    GIT_NO_LAZY_FETCH: "1",
    GIT_EXEC_PATH: undefined,
  };
}
